### PR TITLE
Draft "Use multi-stage Docker builds" and "Prefer smaller base images"

### PR DESCRIPTION
**Pull request by BrunoScheufler**

### CODE REVIEW

File: sections/docker/multi_stage_builds.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
+# Use multi-stage builds
+
+### One Paragraph Explainer
+
+Multi-stage builds allow to separate build- and runtime-specific environment details, such as available binaries, exposed environment variables, and even the underlying operating system.
+Splitting up your Dockerfiles into multiple stages will help to reduce final image and container size as you'll only ship what you really need to run your application. Sometimes, you'll need to
+include tools that are only needed during the build phase, for example development dependencies such as the TypeScript CLI. You can install it during the build stage and only use the final output in the run stage.
```
Comment (kevynb) [inline sections/docker/multi_stage_builds.md:7]: Same thing here, it seems there are linebreaks in the middle of sentences in this block

File: sections/docker/multi_stage_builds.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+- src/
+    - index.ts
+- package.json
+- yarn.lock
+- .dockerignore
+- docs/
+  - README.md
+```
+
+Your .dockerignore will already filter out files that won't be needed for building and running your application.
```
Comment (kevynb) [inline sections/docker/multi_stage_builds.md:26]: We might add a link here to the `.dockerignore` bullet point 🤔

File: sections/docker/multi_stage_builds.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+# Set non-root user and expose port 8080
+USER node
+EXPOSE 8080
+
+WORKDIR /home/node/app
+
+# Copy dependency information and install production-only dependencies
+COPY --chown=node:node package.json yarn.lock ./
+RUN yarn install --production
```
Comment (kevynb) [inline sections/docker/multi_stage_builds.md:69]: we could add `--frozen-lockfile` to ensure that the build uses the packages we used in `dev`

File: sections/docker/smaller_base_images.md [hunk 4]
```diff
@@ -0,0 +1,13 @@
+# Prefer smaller Docker base images
+
+Large Docker images can lead to higher exposure to vulnerabilities and increased resource consumption. Often, you don't need certain packages installed at runtime that are needed for building.
+Pulling and storing larger images will become more expensive at scale, when dealing with larger images. Minimal images may not come with common libraries needed for building native modules or packages
+useful for debugging (e.g. curl) pre-installed, by design. Using the Alpine Linux variants of images, can lead to a reduced footprint in terms of resources used and the amount of attack vectors
+present in fully-featured systems. The Node.js v14.4.0 Docker image is ~345MB in size versus ~39MB for the Alpine version.
```
Comment (kevynb) [inline sections/docker/smaller_base_images.md:6]: Maybe add a statement like "That's almost a 10x reduction" for more impact?

File: README.md [hunk 5]
```diff
@@ -1151,13 +1151,14 @@ Bear in mind that with the introduction of the new V8 engine alongside the new E
 
 <br /><br /><br />
 
-## ![✔] 8.8. Use multistage builds
+## ![✔] 8.8. Use multi-stage builds
```
Comment (goldbergyoni) [inline README.md:1154]: I believe that every reader wishes to know WHY soon. Without this, how can I tell if this topic is of interest? For example, 'Use multi-stage builds to producer leaner and safer images'

cc @js-kyle

File: README.md [hunk 6]
```diff
@@ -1151,13 +1151,14 @@ Bear in mind that with the introduction of the new V8 engine alongside the new E
...
 
-**TL;DR:**
+**TL;DR:** A lot of build-time dependencies and files are not needed for running your application. With multi-stage builds, these resources can be used during
+build, while the runtime environment contains only what's necessary.
```
Comment (goldbergyoni) [inline README.md:1157]: Again, I would say something about WHY is this important. This is also called the 'so what' part

File: README.md [hunk 7]
```diff
@@ -1171,13 +1172,14 @@ Bear in mind that with the introduction of the new V8 engine alongside the new E
 
 <br /><br /><br />
 
-## ![✔] 8.10. Prefer smaller images
+## ![✔] 8.10. Prefer smaller Docker base images
```
Comment (goldbergyoni) [inline README.md:1175]: Consider mentioning also Slim which is popular like Alpine. They don't mitigate these issues as they also have CVEs but they reduce the likelihood of vulnerabilities and compact the attack surface

File: sections/docker/multi_stage_builds.md (hunk 3 above)
Comment (goldbergyoni) [reply sections/docker/multi_stage_builds.md:69]: Does yarn have the equivalent of 'npm ci'?

File: sections/docker/multi_stage_builds.md [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+RUN yarn
+
+# Copy source code (and all other relevant files)
+COPY --chown=node:node src ./src
+
+# Build code
+RUN yarn build
+
+# Run-time stage
+FROM node:14.4.0-alpine
```
Comment (goldbergyoni) [inline sections/docker/multi_stage_builds.md:59]: How is this differerent from copying directly from 'build' (not mentioning the OS again)? Isn't this more intuitive?

@kevynb @BrunoScheufler

File: sections/docker/multi_stage_builds.md [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+EXPOSE 8080
+
+WORKDIR /home/node/app
+
+# Copy dependency information and install production-only dependencies
+COPY --chown=node:node package.json yarn.lock ./
+RUN yarn install --production
+
+# Copy results from previous stage
+COPY --chown=node:node --from=build /home/node/app/dist ./dist
```
Comment (goldbergyoni) [inline sections/docker/multi_stage_builds.md:72]: Can't we just do 'COPY ./dist ./dist'?

File: sections/docker/smaller_base_images.md [hunk 10]
```diff
@@ -0,0 +1,13 @@
+# Prefer smaller Docker base images
+
+Large Docker images can lead to higher exposure to vulnerabilities and increased resource consumption. Often, you don't need certain packages installed at runtime that are needed for building.
```
Comment (goldbergyoni) [inline sections/docker/smaller_base_images.md:3]: There are many line breaks, only to me?

File: sections/docker/smaller_base_images.md (hunk 4 above)
Comment (goldbergyoni) [reply sections/docker/smaller_base_images.md:6]: @BrunoScheufler Consider also including 'Slim' which is very popular. Maybe make the point that working with ~500MB images will make it feasible to push changes only from high-bandwidth lines (happened to me a lot that I couldn't push in planes or trains)

File: README.md [hunk 11]
```diff
@@ -1171,13 +1172,14 @@ Bear in mind that with the introduction of the new V8 engine alongside the new E
...
 
-**TL;DR:**
+**TL;DR:** Large images lead to higher exposure to vulnerabilities and increased resource consumption. Using leaner Docker images, such as Alpine Linux variants,
+mitigates this issue.
```
[the remark this answers is not in the document]
Comment (BrunoScheufler) [reply README.md:1178]: Huh, you're right, will fix that

File: sections/docker/multi_stage_builds.md (hunk 9 above)
Comment (BrunoScheufler) [reply sections/docker/multi_stage_builds.md:72]: You need to copy from the previous step, which is the build stage

File: sections/docker/multi_stage_builds.md (hunk 8 above)
Comment (BrunoScheufler) [reply sections/docker/multi_stage_builds.md:59]: Do you mean using another stage?

**Pull request merged**